### PR TITLE
Refactor SQL update API 

### PR DESCRIPTION
**Pull request by anupama-pathirage**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/14024
>This PR will remvoe `updateWithGeneratedKeys` remote function and that functionality is supported via `update` remote function. Earlier we added two functions as we didn't have support for optional parameters and records in the language. 

**Current API for update operations:**
```ballerina
public remote function update(@sensitive string sqlQuery, Param... parameters) returns int|error {
        return nativeUpdate(self, sqlQuery, ...parameters);
}

public remote function updateWithGeneratedKeys(@sensitive string sqlQuery, string[]? keyColumns,
       Param... parameters) returns (int, string[])|error {
        return nativeUpdateWithGeneratedKeys(self, sqlQuery,keyColumns, ...parameters);
}
```

**Proposed API with this PR:**
```ballerina
# UpdateResult represents the output of the `update` remote function.
#
# + updatedRowCount - The updated row count during the sql statement exectuion
# + generatedKeys - A map of auto generated key values during the sql statement execution
public type UpdateResult record {
    int updatedRowCount;
    map<any> generatedKeys;
    !...;
};

public remote function update(@sensitive string sqlQuery, string[]? keyColumns = (), Param... parameters)
                               returns UpdateResult|error {
        return nativeUpdate(self, sqlQuery, keyColumns = keyColumns, ...parameters);
}
```

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 1]
```diff
@@ -166,23 +168,20 @@ function testGeneratedKeyOnInsertEmptyResults() returns (int|string) {
...
+    } else if (x is error) {
         returnVal = x.reason();
     }
     testDB.stop();
     return returnVal;
 }
 
 
-function testGeneratedKeyWithColumn() returns string {
+function testGeneratedKeyWithColumn() returns (int) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
function testGeneratedKeyWithColumn() returns int {
```

File: stdlib/database/h2/src/test/resources/test-src/h2/h2_actions_test.bal [hunk 2]
```diff
@@ -105,7 +105,7 @@ function testCall() returns (string) {
     return name;
 }
 
-function testGeneratedKeyOnInsert() returns (string) {
+function testGeneratedKeyOnInsert() returns (string|int) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/h2/h2_actions_test.bal]:
```suggestion
function testGeneratedKeyOnInsert() returns string|int {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 3]
```diff
@@ -122,37 +128,33 @@ function testUpdateTableData() returns int {
             password: "",
             poolOptions: { maximumPoolSize: 1 }
         });
-
+    int updateCount = 0;
     var result = testDB->update("Update Customers set country = 'UK' where registrationID = 1");
-    int updateCount = getIntResult(result);
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 4]
```diff
@@ -109,7 +112,10 @@ function testCreateTable() returns int {
             poolOptions: { maximumPoolSize: 1 }
         });
     var result = testDB->update("CREATE TABLE IF NOT EXISTS Students(studentID int, LastName varchar(255))");
-    int returnValue = getIntResult(result);
+    int returnValue = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 5]
```diff
@@ -95,7 +95,10 @@ function testInsertTableData() returns int {
         });
     var result = testDB->update("Insert into Customers (firstName,lastName,registrationID,creditLimit,country)
                                          values ('James', 'Clerk', 3, 5000.75, 'USA')");
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 6]
```diff
@@ -308,7 +299,10 @@ function testInsertTableDataWithParameters() returns int {
 
     var result = testDB->update("Insert into Customers (firstName,lastName,registrationID,creditLimit,country)
                                      values (?,?,?,?,?)", para1, para2, para3, para4, para5);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 7]
```diff
@@ -324,7 +318,10 @@ function testInsertTableDataWithParameters2() returns int {
 
     var result = testDB->update("Insert into Customers (firstName,lastName,registrationID,creditLimit,country)
                                      values (?,?,?,?,?)", "Anne", "James", 3, 5000.75, "UK");
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 8]
```diff
@@ -341,7 +338,10 @@ function testInsertTableDataWithParameters3() returns int {
     string s1 = "Anne";
     var result = testDB->update("Insert into Customers (firstName,lastName,registrationID,creditLimit,country)
                                      values (?,?,?,?,?)", s1, "James", 3, 5000.75, "UK");
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 9]
```diff
@@ -473,7 +473,10 @@ function testINParameters() returns int {
             smallint_type, clob_type, binary_type) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
         paraID, paraInt, paraLong, paraFloat, paraDouble, paraBool, paraString, paraNumeric,
         paraDecimal, paraReal, paraTinyInt, paraSmallInt, paraClob, paraBinary);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 10]
```diff
@@ -494,7 +497,10 @@ function testBlobInParameter() returns (int, byte[]) {
 
     var result = testDB->update("INSERT INTO BlobTable (row_id,blob_type) VALUES (?,?)",
         paraID, paraBlob);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 11]
```diff
@@ -520,7 +526,10 @@ function testINParametersWithDirectValues() returns (int, int, float, float, boo
     var result = testDB->update("INSERT INTO DataTypeTable (row_id, int_type, long_type, float_type,
         double_type, boolean_type, string_type, numeric_type, decimal_type, real_type) VALUES (?,?,?,?,?,?,?,?,?,?)",
         25, 1, 9223372036854774807, 123.34, 2139095039.1, true, "Hello", 1234.567, 1234.567, 1234.567);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 12]
```diff
@@ -577,7 +586,10 @@ function testINParametersWithDirectVariables() returns (int, int, float,
             float_type, double_type, boolean_type, string_type, numeric_type, decimal_type, real_type)
             VALUES (?,?,?,?,?,?,?,?,?,?)", rowid, intType, longType, floatType, doubleType, boolType,
             stringType, numericType, decimalType, realType);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 13]
```diff
@@ -639,7 +651,10 @@ function testNullINParameterValues() returns int {
             smallint_type, clob_type, binary_type) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
         paraID, paraInt, paraLong, paraFloat, paraDouble, paraBool, paraString, paraNumeric,
         paraDecimal, paraReal, paraTinyInt, paraSmallInt, paraClob, paraBinary);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 14]
```diff
@@ -658,7 +673,10 @@ function testNullINParameterBlobValue() returns int {
 
     var result = testDB->update("INSERT INTO BlobTable (row_id, blob_type) VALUES (?,?)",
         paraID, paraBlob);
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 15]
```diff
@@ -672,7 +690,10 @@ function testEmptySQLType() returns int {
             poolOptions: { maximumPoolSize: 1 }
         });
     var result = testDB->update("Insert into Customers (firstName) values (?)", "Anne");
-    int insertCount = getIntResult(result);
+    int insertCount = 0;
+    if(result is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 16]
```diff
@@ -851,7 +872,10 @@ function testDateTimeInParameters() returns int[] {
     sql:Parameter para5 = { sqlType: sql:TYPE_DATETIME, value: "2017-01-30T13:27:01.999999Z" };
 
     var result1 = testDB->update(stmt, para1, para2, para3, para4, para5);
-    int insertCount1 = getIntResult(result1);
+    int insertCount1 = 0;
+    if(result1 is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result1 is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 17]
```diff
@@ -862,7 +886,10 @@ function testDateTimeInParameters() returns int[] {
     para5 = { sqlType: sql:TYPE_DATETIME, value: "-2017-01-30T13:27:01.999999-08:30" };
 
     var result2 = testDB->update(stmt, para1, para2, para3, para4, para5);
-    int insertCount2 = getIntResult(result2);
+    int insertCount2 = 0;
+    if(result2 is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result2 is sql:Result) {
```

File: stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal [hunk 18]
```diff
@@ -874,7 +901,10 @@ function testDateTimeInParameters() returns int[] {
     para5 = { sqlType: sql:TYPE_DATETIME, value: timeNow };
 
     var result3 = testDB->update(stmt, para1, para2, para3, para4, para5);
-    int insertCount3 = getIntResult(result3);
+    int insertCount3 = 0;
+    if(result3 is sql:Result) {
```
Comment (Shan1024) [inline stdlib/database/h2/src/test/resources/test-src/sql/sql_actions_test.bal]:
```suggestion
    if (result3 is sql:Result) {
```

File: stdlib/database/sql/src/main/ballerina/sql/types.bal [hunk 19]
```diff
@@ -142,6 +142,17 @@ public type Parameter record {
...
 
+# Result represents the output of the `update` remote function.
+#
+# + updatedRowCount - The updated row count during the sql statement exectuion
+# + generatedKeys - A map of auto generated key values during the sql statement execution
+#
+public type Result record {
+    int updatedRowCount;
+    map<any> generatedKeys;
+    !...
```
Comment (Shan1024) [inline stdlib/database/sql/src/main/ballerina/sql/types.bal]:
```suggestion
    !...;
```

File: stdlib/database/sql/src/main/ballerina/sql/types.bal [hunk 20]
```diff
@@ -142,6 +142,17 @@ public type Parameter record {
     !...;
 };
 
+# Result represents the output of the `update` remote function.
+#
+# + updatedRowCount - The updated row count during the sql statement exectuion
+# + generatedKeys - A map of auto generated key values during the sql statement execution
+#
```
Comment (Shan1024) [inline stdlib/database/sql/src/main/ballerina/sql/types.bal]: I think we can remove this empty doc line.

File: stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/Constants.java [hunk 21]
```diff
@@ -180,18 +180,29 @@
         public static final String EQUAL_SYMBOL = "=";
     }
 
+    /**
+     * Constants used when appending JDBC driver properties for the connection URL.
+     */
+    public static final class DatabaseProductNames {
```
Comment (Shan1024) [inline stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/Constants.java]: Can't we simply use `Database` or `DatabaseName` here? Having `Product` in the class name does not seem right.

File: stdlib/database/sql/src/main/ballerina/sql/types.bal [hunk 22]
```diff
@@ -142,6 +142,16 @@ public type Parameter record {
     !...;
 };
 
+# Result represents the output of the `update` remote function.
+#
+# + updatedRowCount - The updated row count during the sql statement exectuion
+# + generatedKeys - A map of auto generated key values during the sql statement execution
+public type Result record {
+    int updatedRowCount;
+    map<any> generatedKeys;
```
Comment (Manuri) [inline stdlib/database/sql/src/main/ballerina/sql/types.bal]: Should this be `anydata`?

File: stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/actions/AbstractSQLAction.java [hunk 23]
```diff
@@ -1004,4 +973,15 @@ private int getParameterDirection(BMap<String, BValue> parameter) {
         }
         return direction;
     }
+
+    private static BMap<String, BValue> createResultRecord(Context context, int count, BMap keyValues) {
+        PackageInfo sqlPackageInfo = context.getProgramFile().getPackageInfo(Constants.SQL_PACKAGE_PATH);
+        if (sqlPackageInfo == null) {
+            return null;
```
Comment (Manuri) [inline stdlib/database/sql/src/main/java/org/ballerinalang/database/sql/actions/AbstractSQLAction.java]: What is the scenario that `sqlPackageInfo` becomes `null`?

**Pull request merged**